corehq/apps/hqcase/tasks.py

```diff
@@ -1,93 +1,145 @@
+    restore_user = get_restore_user(domain, couch_user, None)
+    device = MockDevice(restore_user.project, restore_user, {"overwrite_cache": True})
+    sync_result = device.sync()
+
+    cases = {}
+    new_case_ids = {}
+    for case_id, case in sync_result.cases.iteritems():
+        if case.case_type != "commcare-user":
+            cases[case_id] = case
+            new_case_ids[case_id] = [str(uuid.uuid4()) for _ in range(factor - 1)]
+
+    total_cases = len(cases) * (factor - 1)
+    if task:
+        DownloadBase.set_progress(explode_case_task, 0, total_cases)
+
+    queue = []
+    progress = 0
+
+    def queue_case(new_case, queue, progress):
+        queue.append(new_case)
+        if len(queue) >= 500:   # submit 500 cases at a time
+            submit_case_blocks(queue, domain, user_id=user_id, device_id="explode_cases")
+            progress += len(queue)
             if task:
-                DownloadBase.set_progress(explode_case_task, count, 0)
+                DownloadBase.set_progress(explode_case_task, progress, total_cases)
+            del queue[:]
+        return progress
+
+    graph = CaseGraph(cases)
+    for old_case_id in graph.topological_sort():
+        for explosion in range(factor - 1):
+            new_case = copy(cases[old_case_id])
 
-    messages.append("All of %s's cases were exploded by a factor of %d" % (user.raw_username, factor))
+            new_case.create = True
+            new_case.case_id = new_case_ids[old_case_id][explosion]
+            new_case.update['cc_exploded_from'] = old_case_id
+            new_case.update['cc_explosion_id'] = explosion_id
+            new_case.index = {
+                key: IndexAttrs(
+                    i.case_type, new_case_ids[i.case_id][explosion], i.relationship
+                ) for key, i in cases[old_case_id].index.iteritems()
+            }
+            progress += queue_case(ElementTree.tostring(new_case.as_xml()), queue, progress)
 
-    return {'messages': messages}
+    if len(queue):
+        submit_case_blocks(queue, domain, user_id=user_id, device_id="explode_cases")
+        DownloadBase.set_progress(explode_case_task, total_cases, total_cases)
 
+    return {'messages': [
+        "Successfully created {} cases".format(total_cases),
+        "Your explosion_id for this explosion is {}".format(explosion_id)
+    ]}
+
+
+class CaseGraph(object):
+    def __init__(self, cases):
+        self.graph = {}         # case indices
+        self.inverse_graph = defaultdict(list)  # reverse indices
+        self.roots = []
+
+        for case_id, case in cases.iteritems():
+            indices = [idx.case_id for idx in case.index.itervalues()]
+            self.graph[case.case_id] = indices
+            for index in indices:
+                self.inverse_graph[index].append(case.case_id)
+
+            if not indices:
+                self.roots.append(case.case_id)
+
+    def topological_sort(self):
```

Comment:
will this be deterministic where there are multiple cases on the same level? Think the values in the inverse graph should be sorted.